editor-server/src/routes/api/uploadData.ts

```diff
@@ -56,144 +58,190 @@ const uploadData = async (req: Request, res: Response) => {
-            status.length
-          }, Expected ${Object.keys(allPartsList).length}`
-        );
+      if (typeof start !== "number") error.push(`CTRLFRAME_DATA_ERROR: Control frameID ${frameId} has incorrect "start" type. Expected type number, but received ${typeof start}`);
+      if (typeof fade !== "boolean") error.push(`CTRLFRAME_DATA_ERROR: Control frameID ${frameId} has incorrect "fade" type. Expected type boolean, but received ${typeof fade}`);
+      if (!isArray(status)) {
+        error.push(`CTRLFRAME_DATA_ERROR: Control frameID ${frameId} has incorrect "status" type. Expected type array`);
+        return;
       }
-      for (let i = 0; i < status.length; i++) {
-        if (
-          status[i].length !==
-          Object.values(allPartsList[Object.keys(allPartsList)[i]]).length
-        ) {
-          throw new Error(
-            `CTRLFRAME_DATA_ERROR: Control Frame starting at ${start}ms, dancer '${
-              Object.keys(allPartsList)[i]
-            }' has invalid number of parts. Found ${
-              status[i].length
+      if (status.length !== Object.keys(allPartsList).length) error.push(`CTRLFRAME_DATA_ERROR: Control frame starting at ${start}ms has invalid number of dancers. Found ${status.length}, Expected ${Object.keys(allPartsList).length}`);
+      status.forEach((dancerStatus: (TELControl | TLEDControl | TFiberControl)[], dancerIdx) => {
+        if (!isArray(dancerStatus)) {
+          error.push(`CTRLFRAME_DATA_ERROR: Control frameID ${frameId}, status idx ${dancerIdx} has incorrect type. Expected type array`);
+          return;
+        }
+        if (dancerStatus.length !== Object.values(allPartsList[Object.keys(allPartsList)[dancerIdx]]).length) {
+          error.push(
+            `CTRLFRAME_DATA_ERROR: Control frameID ${frameId}, status idx ${dancerIdx} has invalid number of parts. Found ${
+              dancerStatus.length
             }, Expected ${Object.values(allPartsList).length}`
           );
+          return;
         }
-      }
-    });
-
-    // validate status data & fiber color
-    const sortedDancerTmp = Object.keys(allPartsList).sort();
-    Object.values(control).map((frameObj: CtrlFrameTmpData) => {
-      const { fade, start, status } = frameObj;
 
-      for (let i = 0; i < status.length; i++) {
-        for (let j = 0; j < status[i].length; j++) {
-          const tmpPart = allPartsList[sortedDancerTmp[i]][j];
+        // validate status data & fiber color
+        dancerStatus.forEach((partStatus: TELControl | TLEDControl | TFiberControl, partIdx)=>{
+          const tmpPart = dancer[dancerIdx].parts[partIdx];
+          if (!isArray(partStatus)) {
+            error.push(`CTRLFRAME_DATA_ERROR: Control frameID ${frameId}, status idx ${dancerIdx}, part Idx ${partIdx} has incorrect type. Expected type array`);
+            return;
+          }
+          if (partStatus.length != 2) {
```

Comment:
type 'EL' could have length 1

Reply:
EL will not be used this year, we can ignore it for this time